### PR TITLE
Integrated condition variable into scheduler and changed command queue implementation

### LINKED ISSUE #218: Improve scheduler performance

### PR DESCRIPTION
**Pull request by janschopohl**
closes #218

### CODE REVIEW

**Comment by janschopohl** [comment]
Temporarily closed as I had a new idea how to improve this by syncing the ui.

File: source/core/scheduler.cpp [hunk 1]
```diff
@@ -42,15 +50,19 @@ void Scheduler::_run() {
...
-    Task task;
-    if (_taskQueue.pop(task)) {
-      // there was something in the queue, ececute the task
-      task();
-
-    } else {
-      // the queue is empty, sleep for a bit
-      std::this_thread::sleep_for(_sleepMilliseconds);
+    std::unique_lock<std::mutex> lock(_mutex);
+    while (_taskQueue.empty()) {
```
Comment (goldsborough) [inline source/core/scheduler.cpp]: Couldn't we use the predicate feature of the condition variable here?
```
_conditionVariable.wait(lock, [&_taskQueue]{ return !_taskQueue.empty(); })
```

File: source/core/scheduler.cpp [hunk 2]
```diff
@@ -42,15 +50,19 @@ void Scheduler::_run() {
...
+      // wait while there is nothing in the queue, guards for spurious wakeups.
+      _conditionVariable.wait(lock);
     }
+    // make sure the mutex is locked and pop an element from the queue (which is
+    // guaranteed to not be empty)
+    if (!lock) lock.lock();
```
Comment (goldsborough) [inline source/core/scheduler.cpp]: You make these checks here (and in other places), but one of the properties of the condition variable is that the mutex is reacquired upon wakeup. As such I think this check is unnecessary?

File: tests/core/project-test.cpp [hunk 3]
```diff
@@ -89,6 +89,9 @@ class ProjectTestFixture : public ::testing::Test {
     MemoryManager memoryManager = projectModule.getMemoryManager();
     memoryManager.setErrorCallback(errorCallback);
     architectureValidator.validate();
+    CommandInterface commandInterface = projectModule.getCommandInterface();
+    commandInterface.setSyncCallback(
+        [this]() { this->projectModule.guiReady(); });
```
Comment (goldsborough) [inline tests/core/project-test.cpp]: There's special syntax that let's you write a lambda without arguments simply as `[...]{...}` (no parantheses)

**Review by goldsborough — CHANGES_REQUESTED** [review]

**Pull request merged**